### ISSUE TITLE
Why does casting result in cross joins?

### ISSUE THREAD

**Issue by jortiz16** [uwescience/raco#419]
This is more of a question, but I'm not sure if this is a result of a bug. I have a dataset in which I ingested with the wrong datatypes for one of the columns. So as a result, I was forced to cast the column when trying to run a join query. 

I don't see why casting resulted in a different plan for my query. When I cast, one of my tables is broadcast to each node and a cross join is done. Whereas if no casting is done, each table is shuffled before a symmetric hash join runs. 

To show this example, I'll use TwitterK as an example. 

Regular query:

```
T1 = scan(TwitterK);
T2 = scan(TwitterK);
Joined = [from T1, T2
          where T1.$1 = T2.$0
          emit T1.$0 as src, T1.$1 as link, T2.$1 as dst];
store(Joined, TwoHopsInTwitter);
```

Query Plan:

<img src="https://cloud.githubusercontent.com/assets/3239603/7558870/432540da-f762-11e4-9f35-48e5675895c4.png" width="400">

This is the query with casting:
```
T1 = scan(TwitterK);
T2 = scan(TwitterK);
Joined = [from T1, T2
          where int(T1.$1) = int(T2.$0)
          emit T1.$0 as src, T1.$1 as link, T2.$1 as dst];
store(Joined, TwoHopsInTwitter);
```

Query Plan:

<img src="https://cloud.githubusercontent.com/assets/3239603/7558875/58c0775c-f762-11e4-8a4d-4bd3f32f5a6b.png" width="400">

**Comment by senderista** [comment]
I'm not sure how Myria could do any better than a cross join + selection for equijoin on arbitrary predicates? If you cast the attributes separately from the join, you get the query plan you would expect:
```
T1 = scan(TwitterK);
T2 = scan(TwitterK);
T1_cast = [from T1 emit int($0), int($1), int($2)];
T2_cast = [from T2 emit int($0), int($1), int($2)];
Joined = [from T1_cast, T2_cast
          where T1_cast.$1 = T2_cast.$0
          emit T1_cast.$0 as src, T1_cast.$1 as link, T2_cast.$1 as dst];
store(Joined, TwoHopsInTwitter);
```
![image 2015-05-11 at 10 35 32 am](https://cloud.githubusercontent.com/assets/131425/7571088/f5ccc8ae-f7c9-11e4-81ba-8505f0cc81ae.png)

```
MyriaStore(public:adhoc:TwoHopsInTwitter)[MyriaApply(src=$0,link=$1,dst=$2)[MyriaSplitConsumer[MyriaSplitProducer[MyriaSymmetricHashJoin(($1 = $2); $0,$1,$3)[MyriaShuffleConsumer[MyriaShuffleProducer(h($1))[MyriaApply(_COLUMN0_=CAST(LONG_TYPE, $0),_COLUMN1_=CAST(LONG_TYPE, $1))[MyriaScan(public:adhoc:TwitterK)]]],MyriaShuffleConsumer[MyriaShuffleProducer(h($0))[MyriaApply(_COLUMN0_=CAST(LONG_TYPE, $0),_COLUMN1_=CAST(LONG_TYPE, $1))[MyriaScan(public:adhoc:TwitterK)]]]]]]]]
```

**Comment by senderista** [comment]
It occurred to me that we could possibly handle the cross-product case more efficiently if, instead of always broadcasting one relation to each partition of the other relation, we used a variant of 2D hypercube shuffle, where each relation is partitioned in a completely arbitrary way (as long as skew isn't excessive), and we evaluate the symmetric hash join locally on each pair of partitions. This should be better than asymmetric broadcast when both relations are very large. Feel free to create a separate issue for this optimization.

**Comment by jortiz16** [comment]
Thanks @senderista ! I agree and what you mentioned was exactly the problem I had. The tables I was joining are so large that a cross join would quickly cause an out of memory error in Myria. @lbrendanl and I ended up re-ingesting the data with correct data types to prevent casting altogether (although your temporary fix from earlier should work). Given your suggestion, does this issue belong here in Raco, or is it more on the Myria side?

**Comment by bmyerz** [comment]
@jortiz16 I think the example you give is best improved in Raco. Raco
should be creating Tobin's plan for both queries. We just need a way to
recognize more equijoins. I suspect that the crossproduct+select =>
equijoin Rule is not firing in the casting query because our firing rule is
not smart enough. Create an issue? If you'd like to add this rule, I'm
happy to help

**Comment by senderista** [comment]
I just discussed my proposed distributed cross-product with @mbalazin and @jingjingwang and they pointed out that it would result in more data movement with the same amount of data being processed locally, relative to the broadcast join currently used. The only issue with the current cross-product implementation in MyriaX is that we should be using an asymmetric hash join in cases where the broadcast relation is much larger than a partition of the partitioned relation, rather than a symmetric hash join. In other words, we should finish building a hash table on the partitioned relation and then do a streaming probe of this hash table with the broadcast relation. This will completely avoid materializing any tuples in the broadcast relation, although it blocks on building the hash table, so it is not fully pipelined like symmetric hash join.

**Comment by senderista** [comment]
Opened #422 for the cross-product optimization.

**Comment by bmyerz** [comment]
To make it clear: this issue will be resolved by adding the rule that still generates hash join when there is casting.

**Comment by senderista** [comment]
Yes, I believe that raco should factor out expressions in the join predicate into an apply like I did manually above. There's no need to implement the optimization in #419 here, since the cross product would be avoided anyway.

**Comment by bmyerz** [comment]
Then I think you've answered my open question in #422 then

**Comment by senderista** [comment]
Well, if we still have to deal with cross products in some contexts (perhaps where we want some subset of the cartesian product that cannot be expressed as a join), then I think #422 is worth implementing, but if we can avoid cross products in the most common scenarios then I don't think it should be very high priority.

**Comment by bmyerz** [comment]
We'll keep it closed until we take care of the low hanging fruit and we still have a problem
